File: rack/drivers/camera_proxy.cpp
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2006 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Marko Reimer     <reimer@l3s.de>
 *      Joerg Langenberg <joerg.langenberg@gmx.net>
 *
 */
#include <drivers/camera_proxy.h>

//
// proxy functions
//

int CameraProxy::getData(camera_data *recv_data, ssize_t recv_datalen,
                          RACK_TIME timeStamp, uint64_t reply_timeout_ns)
{
    MessageInfo msgInfo;
    
    int ret = RackDataProxy::getData((void *)recv_data, recv_datalen, timeStamp,
                                    reply_timeout_ns, &msgInfo);
    if (ret)
    {
        return ret;
    }

    recv_data = CameraData::parse(&msgInfo);
    return 0;
}

int CameraProxy::getParam(camera_param_data *recv_data, ssize_t recv_datalen,
                           uint64_t reply_timeout_ns)
{
    MessageInfo msgInfo;
    
    int ret = proxyRecvDataCmd(MSG_CAMERA_GET_PARAMETER, MSG_CAMERA_PARAMETER,
                              (void *)recv_data, recv_datalen,
                              reply_timeout_ns, &msgInfo);
    if (ret)
    {
        return ret;
    }

    recv_data = CameraParamData::parse(&msgInfo);
    return 0;
}

File: rack/drivers/CameraProxy.java
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2006 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Marko Reimer     <reimer@l3s.de>
 *      Joerg Langenberg <joerg.langenberg@gmx.net>
 *
 */
package rack.drivers;

import rack.main.*;
import rack.main.tims.*;

public class CameraProxy extends RackDataProxy {

    public static final byte MSG_CAMERA_GET_PARAMETER =
        RackProxy.MSG_POS_OFFSET + 1;

    public static final byte MSG_CAMERA_SET_FORMAT =
        RackProxy.MSG_POS_OFFSET + 2;

    public static final byte MSG_CAMERA_PARAMETER =
        RackProxy.MSG_NEG_OFFSET - 1;

    public static final byte MSG_CAMERA_FORMAT =
        RackProxy.MSG_NEG_OFFSET - 2;


    public CameraProxy(int id, TimsMbx replyMbx)
    {
        super(RackName.create(RackName.CAMERA, id), replyMbx, 5000, 1000, 5000);
        this.id = id;
    }

    public synchronized CameraDataMsg getData(int recordingTime)
    {
        try
        {
            TimsRawMsg raw = getRawData(recordingTime);

            if(raw != null)
            {
                CameraDataMsg data = new CameraDataMsg(raw);
                return data;
            }
            else
            {
                return null;
            }
        }
        catch(TimsException e)
        {
            System.out.println(e.toString());
            return null;
        }
    }

    public synchronized CameraDataMsg getData()
    {
        try
        {
            TimsRawMsg raw = getNextData();

            if(raw != null)
            {
                CameraDataMsg data = new CameraDataMsg(raw);
                return data;
            }
            else
            {
                return null;
            }
        }
        catch(TimsException e)
        {
            System.out.println(e.toString());
            return null;
        }
    }

    public synchronized void setFormat(CameraFormatMsg format)
    {
        currentSequenceNo++;
        System.out.println(format);

        try {
            replyMbx.send(
                MSG_CAMERA_SET_FORMAT,
                commandMbx,
                (byte)0,
                (byte)currentSequenceNo,
                format);

            TimsRawMsg reply;

            do {
                reply = replyMbx.receive(dataTimeout);
            } while (reply.seqNr != currentSequenceNo);

            if (reply.type == RackProxy.MSG_OK) {
                System.out.println(RackName.nameString(replyMbx.getName()) +
                        ": cameraProxy setFormat");
            } else {
                System.out.println(
                        RackName.nameString(replyMbx.getName())
                        + ": "
                        + RackName.nameString(commandMbx)
                        + ".setFormat replied error");
            }
        } catch (TimsException e) {
            System.out.println(RackName.nameString(replyMbx.getName()) +
                    ": cameraProxy setFormat " + e);
        }
    }

    public int getCommandMbx()
    {
        return(RackName.create(RackName.CAMERA, id));
    }
}
